{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ---- Steps and Template Notes (this notebook) ---- 0000.00.00\n",
    "## See the end of this notebook for R configuration information\n",
    "## This is the outline notebook for templates for performing single cell analysis\n",
    "## This series was recoded to include multiplet discover and removal\n",
    "## the original templates were coded dd.d.d. This set was expanded by adding zeros\n",
    "## so that we get dd00.0d.0d. For the notebook \"---- Partition Cell Set Correlation\n",
    "## Analysis-Network Notebooks C ---- \", was recoded from 06.0.0 to 0650.00.00\n",
    "##\n",
    "## ---- Target Data Prep Note ---- 0100.00.00\n",
    "## Prepare Input and Output Data Directories (R)\n",
    "## Identify and Set Up Analysis Parameters (R)\n",
    "## Store Data and Analsyis Parameters Used for All Notebooks (Parm1)\n",
    "##\n",
    "## ---- Global Cell Set Preprocessing Notebook ---- 0200.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Download Aggregated 10x Single Cell Data (Seurat)\n",
    "## Perform QC Filtering (Seurat)\n",
    "## Identify Patient and Health Condition All Cells\n",
    "## Create Filtered Monocle3 CDS Object and Perform Down Sampling (Monocle3)\n",
    "## Remove Zero Expressing Genes\n",
    "## Perform Preliminary Pre-Processing and Clustering (Monocle3)\n",
    "## Annotate Partitions (Cell Types)\n",
    "## Save Result As An R Data Object (CDS1)\n",
    "##\n",
    "## ---- Global Cell Set Characterization Notebook 15 Partition Cells Set ---- 0240.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Read R Data Object from Prior Notebook (CDS1)\n",
    "## Compute Top Genes for 15 Partition Cell Set and Save Result\n",
    "## Create Dot Plots For 15 Partition Cell Set\n",
    "##\n",
    "## ---- Global Cell Set Characterization Notebook Doublet Identification ---- 0250.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Preprocessed/Downsampled CDS Object\n",
    "## Preform some preliminary doublet analyses (cut and paste outputted to spreadsheets)\n",
    "## Create new remixed partitions\n",
    "## Perform formal Gene Exclusion Study (Look for Genes Common in One Partition and Rare in Others)\n",
    "## Apply Gene Selection Based on Gene Exclusion Study\n",
    "## Perform Doublet (Multiplet) Scoring\n",
    "## Save Result As An R Data Object (CDS1)\n",
    "## Compute the mean of the maximum pseudo_R2rds for Tom's genes in the dot plot\n",
    "##\n",
    "## ---- Global Cell Set Doublet Filtering and Post Processing Notebook ---- 0270.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Preprocessed/Downsampled Doublet Identified CDS Object\n",
    "## Create Various Plots\n",
    "## Apply Doublet Filtering\n",
    "## Reformulate Partitions as Needed (and assigned cell types)\n",
    "## Save Result As An R Data Object (CDS1)\n",
    "##\n",
    "## ---- Global Cell Set Characterization Notebook ---- 0300.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Read R Data Object from Prior Notebook (CDS1)\n",
    "## Characterize Partition - Statatics (Monocle3)\n",
    "## Compute Top Genes By Partition an Save Result\n",
    "## Create Global Cell Plots as Needed and Save Results\n",
    "## Create Global Partition DOT Plots as Needed and Save Results\n",
    "## Perform Eigencell Analysis and Plot Results as Needed and Save Results\n",
    "## Perform Cross Partition Network Analysis\n",
    "## Save CDS Result As An R Data Object (CDS2)\n",
    "##\n",
    "## ---- Global Cell Set Characterization Notebook ---- ---- 0301.00.00\n",
    "## This is a modified version of the 0301.00.00 Notebook\n",
    "## for Some Reformatted Plots Needed for the Paper\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Read R Data Object from Prior Notebook (CDS1)\n",
    "## Characterize Partition - Statatics (Monocle3)\n",
    "## Compute Top Genes By Partition an Save Result (added all 10/29/2019)\n",
    "## Create Global Cell Plots as Needed and Save Results\n",
    "## Create Global Partition DOT Plots as Needed and Save Results\n",
    "## Perform Eigencell Analysis and Plot Results as Needed and Save Results\n",
    "## Perform Cross Partition Network Analysis\n",
    "## Save CDS Result As An R Data Object (CDS2)\n",
    "## Compute the mean of the maximum pseudo_R2rds for Tom's genes in the dot plot\n",
    "## \n",
    "## ---- Partition Cell Set(s) Extraction Notebook ---- 0400.00.00\n",
    "## Read R Data Object from Prior Notebook (CDS1)\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Loop Through Each Partition and Perform Network Analysis as Follows:\n",
    "## Extract Partition Cell Subsets and Save Them\n",
    "## Extract Differential Expression Gene Set Models and Save Them \n",
    "##\n",
    "## ---- Partition Cell Set Analysis with Diff Expression Notebooks A ---- 0500.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Partition Differential Expresssion Gene Set Model\n",
    "## Plot Cell Sets as Needed\n",
    "## Extract Differential Analysis Coefficients and Save\n",
    "## Extract Gene Rank List and Save\n",
    "##\n",
    "## ---- Partition Cell Set Analysis with Diff Expression Notebooks A ---- 0501.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Partition Differential Expresssion Gene Set Model\n",
    "## Plot Cell Sets as Needed, Like in 0500.00.00, but using original UMAP layout\n",
    "##\n",
    "## ---- Partition Cell Set Analysis - Extended Analysis ---- 0530.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Partition Differential Expresssion Gene Set Model\n",
    "## Create Violin Plots Based On p-value Rank Instead of q-value\n",
    "## Perform Top Genes Analysis Per Cluster and Save\n",
    "## Compute Cluster Stats and Save\n",
    "## Perform Other Plots as Needed\n",
    "##\n",
    "## ---- Partition Cell Set Analysis - Extended Analysis ---- 0531.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Partition Differential Expresssion Gene Set Model\n",
    "## Load Top Genes Analysis Per Cluster and Save\n",
    "## Plot Dot Plot by Clusters\n",
    "## Plot Partition Cell Clots Colorized by Patient\n",
    "## Plot Original 15 Partition Cell Set by Colorized by Patient\n",
    "## Plot Original 15 Partition Cell Set by Colorized by Partition and Labelled by Cell Type\n",
    "## Perform Other Plots as Needed\n",
    "##\n",
    "## ---- Partition Cell Set Analysis - Extended Analysis 2 ---- 0550.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Partition Differential Expresssion Gene Set Model\n",
    "## Create Volcano Plots\n",
    "## Create Special Volcano Plots for VSMC and EC Partitions\n",
    "## Create Special Gene/Cell Plots for VSMC and EC Partitions\n",
    "## Perform Other Plots as Needed\n",
    "##\n",
    "## ---- Partition Cell Set Partital Correlation Analysis-Network Notebooks B ---- 0600.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Partition Differential Expresssion Coefficients\n",
    "## Perform Partial Correlation Analysis\n",
    "## Perform Perform FDR Scoring\n",
    "## Create Network with Extreme Cut\n",
    "## Annotate Network with Differential Expression Results\n",
    "## Plot Network as Needed\n",
    "## Save Tabulated and Summarized Partition Results (PartSum1)\n",
    "##\n",
    "## ---- Partition Cell Set Partital Correlation Analysis-Network Notebooks B ---- 0601.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Partition Differential Expresssion Coefficients\n",
    "## Perform Partial Correlation Analysis\n",
    "## Perform Perform FDR Scoring\n",
    "## Create Network with Extreme Cut, But With Inverted Weights\n",
    "## Annotate Network with Differential Expression Results\n",
    "## Plot Network as Needed\n",
    "## Save Tabulated and Summarized Partition Results (PartSum1)\n",
    "##\n",
    "## ---- Partition Cell Set Partital Correlation Analysis-Network Notebooks B ---- 0620.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Perform Partial Correlation Analysis\n",
    "## Perform Perform FDR Scoring\n",
    "## Create 50000 Edge Networks and Save\n",
    "## Save Tabulated and Summarized Partition Results\n",
    "##\n",
    "## ---- Partition Cell Set Correlation Analysis-Network Notebooks C ---- 0650.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Partition Differential Expresssion Coefficients\n",
    "## Perform Correlation Analysis\n",
    "## Perform Perform FDR Scoring\n",
    "## Create Network with Extreme Cut\n",
    "## Annotate Network with Differential Expression Results\n",
    "## Plot Network as Needed\n",
    "##\n",
    "## ---- Partition Cell Set Network Module Analysis ---- 0660.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Networks Created in 0600.00.00.Carotid.SC.NotebookPartitionAnalysisNetwork.B.01.\n",
    "## These networks were created with a cutoff based on 1E-1\n",
    "## Plot Dense Networks based on FDR cutoff 1E-1\n",
    "##\n",
    "## ---- Partition Cell Set Network Module Analysis ---- 0670.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Large 50,000 Edge Networks Created in 0620.00.00.Carotid.SC.NotebookPartitionAnalysisNetwork.B.01.ipynb0620.00.00\n",
    "## Cut networks to 20,000 edges\n",
    "## Create Clusters Using Recursive Louvain\n",
    "## Plot the Network with Clusters (20,000 Edges Each)\n",
    "## The following items will be performed in another notebook\n",
    "## Perform Module GO Enrichment Analysis on Selected Clusters (topGO, with Tables and GO Plots)\n",
    "## Note: Use GO Biological Process Terms for Enrichment Analysis\n",
    "## Save Module Analysis Results to 'Modules.20000' Directory \n",
    "##\n",
    "## ---- Partition Cell Set Network Module Analysis ---- 0671.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load Large 50,000 Edge Networks Created in \n",
    "##       0620.00.00.Carotid.SC.NotebookPartitionAnalysisNetwork.B.01.ipynb0620.00.00\n",
    "## Cut networks to 2,000 edges\n",
    "## Create Clusters Using Recursive Louvain\n",
    "## Plot the Network with Clusters (2,000 Edges Each)\n",
    "## The following items will be performed in another notebook\n",
    "## Perform Module GO Enrichment Analysis on Selected Clusters (topGO, with Tables and GO Plots)\n",
    "## Note: Use GO Biological Process Terms for Enrichment Analysis\n",
    "## Save Module Analysis Results to 'Modules.2000' Directory \n",
    "##\n",
    "## ---- Partition Cell Set Network Module Analysis ---- 0675.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load 20K Data Sets Created in 0670.00.00.Carotid.SC.NotebookPartition.Network.Modules.01\n",
    "## Perform Module DE Analysis on Selected Clusters\n",
    "## Create Module and Gene Tabulated Files, Also For IPA\n",
    "## Find Network Stats for earch Partition (FDR Thresholds)\n",
    "##\n",
    "## ---- Partition Cell Set Network Module Analysis ---- 0676.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load 2K Data Sets Created in 0671.00.00.Carotid.SC.NotebookPartition.Network.Modules.01\n",
    "## Perform Module DE Analysis on Selected Clusters\n",
    "##\n",
    "## ---- Partition Cell Set Network Module Analysis ---- 0677.00.00 - Outdated Module Network Plot, H5 Data Example\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load 20K Data Sets Created in 0670.00.00.Carotid.SC.NotebookPartition.Network.Modules.01\n",
    "## Create Special Network/Module Plots for VSMC and EC Partitions\n",
    "## Some Minor Preparation for Using tkplot in R-Studio\n",
    "## Some Preparation That Will Aid in Assessing Read/UMI Distributions (Molecule_info 10X h5 file)\n",
    "##\n",
    "## ---- Partition Cell Set Network Module Analysis ---- 0678.00.00 - Outdated Module Network Plot\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Load 20K Data Sets Created in 0670.00.00.Carotid.SC.NotebookPartition.Network.Modules.01\n",
    "## Create Special Network/Module Plots for VSMC and EC Partitions (Using Degree Connectivity and selected genes)\n",
    "## Plot Modules Using Plot Results from tkplot Run From R-Studio\n",
    "##\n",
    "## ---- Partition Cell Set Network Module Analysis ---- 0680.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Plot Early Attempts at Heatmaps, Experimenting with Format\n",
    "##\n",
    "## ---- Partition Cell Set Network Heatmap Module Analysis ---- 0681.01.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "##\n",
    "## ---- Partition Cell Set Network Hallmark Analysis ---- 0681.02.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Plot Hallmark Gene Heatmaps\n",
    "## Create DE Analysis on VSCM Hallmark Subgroup Genes\n",
    "## Create Volcano Plot for VSMC DE Hallmark Subgroup Genes\n",
    "##\n",
    "## ---- Partition Cell Set Network Heatmap Module Analysis ---- 0681.03.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Plot Module Gene Heatmaps\n",
    "## Expermiments in Heat Map Differential Expression Analysis, Using\n",
    "## Monocle, GLM, DESeq2, and speedglm\n",
    "##\n",
    "## ---- Partition Cell Set Network Heatmap Module Analysis ---- 0681.04.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Each Partition and Perform Analysis as Follows:\n",
    "## Load Partition Cell Set\n",
    "## Plot Module Gene Heatmaps, Amended fro 0681.01.00 to Address Nature Review Comments\n",
    "## Formalized Heatmap Differianl Express Analysis as Started in 0681.03.00\n",
    "##\n",
    "## ---- Partition Cell Set Network Module Cell Plotting ---- 0682.03.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Monocle Cell Sets\n",
    "## Perform Special Gene (gene enrichment) Plots, Using Customized Code\n",
    "##\n",
    "## ---- Partition Cell Set Network Module Plotting ---- 0682.00.00\n",
    "## Heavily Derived From and Modified From 0678.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Develop and Implement a Comprehensive Network Plotting Strategy\n",
    "## Plot Network Modules Using New Plotting Strategy\n",
    "##\n",
    "## ---- Partition Cell Set Network Module Cell Plotting ---- 0682.01.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Monocle Cell Sets\n",
    "## Perform Special Gene (gene enrichment) Plots\n",
    "##\n",
    "## ---- Partition Cell Set Network Module Plotting ---- 0683.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load Network Utilities\n",
    "## Loop Through Partitions 2 and 3, Loading Medium Density Networks\n",
    "## Loop Through Partitions 2 and 3, Loading High Density Networks\n",
    "## Use Network Plotting Strategy To Replot Medium and High Density Networks\n",
    "##\n",
    "## ---- Global Cell Set Doublet Testing with Scrublet Abondoned ---- 00685.00.00\n",
    "## Load aggregated data set (previously unprocessed 10x data)\n",
    "## Perform Scrublet analysis\n",
    "## Save doublet scores, predicted doublets, and cell barcodes\n",
    "##\n",
    "## ---- Global Cell Set Doublet Testing with Scrublet ---- 0686.00.00\n",
    "## This code mirrors the original Scrublet demo, where the full aggregated\n",
    "## Data set was used. Refer to 0686.01.00 for the version that worked with\n",
    "## The downsampled remixed cell set\n",
    "## Load aggregated data set (previously unprocessed 10x data)\n",
    "## Perform Scrublet analysis\n",
    "## Save doublet scores, predicted doublets, and cell barcodes\n",
    "##\n",
    "## ---- Global Cell Set Doublet Testing with Scrublet ---- 0686.01.00\n",
    "## This code is new from 0686.00.00 in that is uses the downsampled\n",
    "## and remixed cell set\n",
    "## Load aggregated data set (previously unprocessed 10x data)\n",
    "## Filter to the downsampled and remixed set (remove QC Failed cells too)\n",
    "## Perform Scrublet analysis\n",
    "## Save doublet scores, predicted doublets, and cell barcodes\n",
    "##\n",
    "## ---- Global Cell Set Doublet Testing with Scrublet ---- 00686.00.00\n",
    "## Load aggregated data set (previously unprocessed 10x data)\n",
    "## Perform Scrublet analysis (try several exprected doublet rates)\n",
    "## Save doublet scores, predicted doublets, and cell barcodes\n",
    "##\n",
    "## ---- Global Cell Set Doublet Testing with DoubletDecon ---- 0687.00.00\n",
    "## This notebook will be abandoned in favor of 0687.02.00 which uses\n",
    "## the downsampled/remixed cell set \n",
    "## Load aggregated data set (previously unprocessed 10x data)\n",
    "## Perform DoubletDecon analysis\n",
    "## Save doublet scores, predicted doublets, and cell barcodes\n",
    "##\n",
    "## ---- Global Cell Set Doublet Testing with DoubletDecon ---- 0687.01.00\n",
    "## Modified to run in X11 session (not really needed, but to see if GUI would be invoked)\n",
    "## Load Seurat PBMC cell set\n",
    "## Perform DoubletDecon analysis\n",
    "## Save doublet scores, predicted doublets, and cell barcodes\n",
    "##\n",
    "## ---- Global Cell Set Doublet Testing with DoubletDecon ---- 0687.02.00\n",
    "## Load aggregated data set (previously unprocessed 10x data)\n",
    "## Filter to downsampled and remixed data sets (remove QC failed cells too)\n",
    "## Perform DoubletDecon analysis\n",
    "## Save doublet scores, predicted doublets, and cell barcodes\n",
    "##\n",
    "## ---- Global Doublelet Methods Comparison---- 0688.00.00\n",
    "## Load Analysis Parameters (Parm1)\n",
    "## Load The Downsampled CDS Object\n",
    "## Load the Scrublet Results (Scores and Marked Cells)\n",
    "## Load the DoubletDecon Results (Doubled Marked Cells)\n",
    "## Load the Gene Exclusion Remix and Doublet Results\n",
    "## Annotate the Three Doublet Methods into the CDS Object\n",
    "## Perform Comparisons of Doublet Methods##\n",
    "##\n",
    "## ---- Utility Code - Alternative Moncole 3 plot_genes_by_group ---- 0700.00.00\n",
    "## Change plot_genes_by_group to plot_genes_by_group2\n",
    "## Change code to allow for as_ordered for pass parameter ordering_type\n",
    "## Convert/Save Notebook at standard r script file\n",
    "##\n",
    "## ---- Pyhton Modules ---- 0701.00.00\n",
    "## Check version of python and modules (if needed) \n",
    "##\n",
    "## ---- Other Utilities - Custering and Plotting From Sara Smith\n",
    "## These May Have Slight Modifications from Sara's Original\n",
    "## /gpfs/home/devans/code/Utilities/fromSara/setup_igraph.r\n",
    "## /gpfs/home/devans/code/Utilities/fromSara/recursive_louvain.r\n",
    "## /gpfs/home/devans/code/Utilities/fromSara/cluster_score.r\n",
    "## /gpfs/home/devans/code/Utilities/fromSara/cluster_overlap.r\n",
    "## /gpfs/home/devans/code/Utilities/fromSara/plot_by_cluster.r\n",
    "##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## R Libraries Used In This Analysis\n",
    "library(biomaRt)\n",
    "library(clusterProfiler)\n",
    "library(corpcor)\n",
    "library(DeconRNASeq)\n",
    "library(doParallel)\n",
    "library(DoubletDecon)\n",
    "library(dplyr)\n",
    "library(EnhancedVolcano)\n",
    "library(fdrtool)\n",
    "library(foreach)\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "library(gplots)\n",
    "library(grid)\n",
    "library(gridExtra)\n",
    "library(grid)\n",
    "library(igraph)\n",
    "library(irlba)\n",
    "library(magrittr)\n",
    "library(Matrix)\n",
    "library(MCL)\n",
    "library(mixOmics)\n",
    "library(monocle3)\n",
    "library(mygene)\n",
    "library(patchwork)\n",
    "library(plyr)\n",
    "library(readxl)\n",
    "library(repr)\n",
    "library(reticulate)\n",
    "library(rhdf5)\n",
    "library(R.utils)\n",
    "library(SDMTools)\n",
    "library(Seurat)\n",
    "library(stringr)\n",
    "library(tidyr)\n",
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 3.5.1 (2018-07-02)\n",
       "Platform: x86_64-pc-linux-gnu (64-bit)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS: /opt/applications/R/3.5.1/gnu/lib64/R/lib/libRblas.so\n",
       "LAPACK: /opt/applications/R/3.5.1/gnu/lib64/R/lib/libRlapack.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.utf8       LC_NUMERIC=C             \n",
       " [3] LC_TIME=en_US.utf8        LC_COLLATE=en_US.utf8    \n",
       " [5] LC_MONETARY=en_US.utf8    LC_MESSAGES=en_US.utf8   \n",
       " [7] LC_PAPER=en_US.utf8       LC_NAME=C                \n",
       " [9] LC_ADDRESS=C              LC_TELEPHONE=C           \n",
       "[11] LC_MEASUREMENT=en_US.utf8 LC_IDENTIFICATION=C      \n",
       "\n",
       "attached base packages:\n",
       " [1] stats4    grid      parallel  stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] forcats_0.5.0               purrr_0.3.4                \n",
       " [3] readr_1.3.1                 tibble_3.0.4               \n",
       " [5] tidyverse_1.3.0             tidyr_1.1.2                \n",
       " [7] stringr_1.4.0               Seurat_3.1.5               \n",
       " [9] SDMTools_1.1-221.1          R.utils_2.9.2              \n",
       "[11] R.oo_1.23.0                 R.methodsS3_1.8.0          \n",
       "[13] rhdf5_2.26.2                reticulate_1.18            \n",
       "[15] repr_1.1.0                  readxl_1.3.1               \n",
       "[17] plyr_1.8.6                  patchwork_1.0.1            \n",
       "[19] mygene_1.18.0               GenomicFeatures_1.34.8     \n",
       "[21] AnnotationDbi_1.44.0        monocle3_0.1.2             \n",
       "[23] SingleCellExperiment_1.4.1  SummarizedExperiment_1.12.0\n",
       "[25] DelayedArray_0.8.0          BiocParallel_1.16.6        \n",
       "[27] matrixStats_0.57.0          GenomicRanges_1.34.0       \n",
       "[29] GenomeInfoDb_1.18.2         IRanges_2.16.0             \n",
       "[31] S4Vectors_0.20.1            mixOmics_6.6.2             \n",
       "[33] lattice_0.20-41             MASS_7.3-51.6              \n",
       "[35] MCL_1.0                     magrittr_1.5               \n",
       "[37] irlba_2.3.3                 Matrix_1.2-18              \n",
       "[39] igraph_1.2.6                gridExtra_2.3              \n",
       "[41] gplots_3.0.4                ggpubr_0.4.0               \n",
       "[43] fdrtool_1.2.15              EnhancedVolcano_1.0.1      \n",
       "[45] ggrepel_0.8.2               dplyr_1.0.2                \n",
       "[47] DoubletDecon_1.1.4          doParallel_1.0.15          \n",
       "[49] iterators_1.0.12            foreach_1.5.0              \n",
       "[51] DeconRNASeq_1.24.0          ggplot2_3.3.2              \n",
       "[53] pcaMethods_1.74.0           Biobase_2.42.0             \n",
       "[55] BiocGenerics_0.28.0         limSolve_1.5.6             \n",
       "[57] corpcor_1.6.9               clusterProfiler_3.10.1     \n",
       "[59] biomaRt_2.38.0             \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] rtracklayer_1.42.2       pbdZMQ_0.3-3             acepack_1.4.1           \n",
       "  [4] bit64_0.9-7              knitr_1.29               data.table_1.13.2       \n",
       "  [7] rpart_4.1-15             RCurl_1.98-1.2           generics_0.1.0          \n",
       " [10] cowplot_1.1.0            RSQLite_2.2.0            RANN_2.6.1              \n",
       " [13] europepmc_0.4            future_1.20.1            chron_2.3-55            \n",
       " [16] bit_1.1-15.2             enrichplot_1.2.0         lubridate_1.7.9         \n",
       " [19] xml2_1.3.2               assertthat_0.2.1         viridis_0.5.1           \n",
       " [22] xfun_0.15                hms_0.5.3                evaluate_0.14           \n",
       " [25] fansi_0.4.1              progress_1.2.2           dbplyr_1.4.4            \n",
       " [28] caTools_1.17.1.2         DBI_1.1.0                htmlwidgets_1.5.2       \n",
       " [31] rARPACK_0.11-0           ellipsis_0.3.1           RSpectra_0.16-0         \n",
       " [34] backports_1.2.0          vctrs_0.3.4              ROCR_1.0-7              \n",
       " [37] abind_1.4-5              withr_2.3.0              ggforce_0.3.2           \n",
       " [40] triebeard_0.3.0          checkmate_2.0.0          sctransform_0.3.1       \n",
       " [43] GenomicAlignments_1.18.1 prettyunits_1.1.1        cluster_2.1.0           \n",
       " [46] DOSE_3.8.2               lazyeval_0.2.2           ape_5.4                 \n",
       " [49] IRdisplay_0.7.0          crayon_1.3.4             ellipse_0.4.2           \n",
       " [52] pkgconfig_2.0.3          tweenr_1.0.1             nlme_3.1-148            \n",
       " [55] nnet_7.3-14              rlang_0.4.8              globals_0.13.1          \n",
       " [58] lifecycle_0.2.0          modelr_0.1.8             rsvd_1.0.3              \n",
       " [61] cellranger_1.1.0         polyclip_1.10-0          lmtest_0.9-38           \n",
       " [64] urltools_1.7.3           IRkernel_1.1             carData_3.0-4           \n",
       " [67] Rhdf5lib_1.4.3           zoo_1.8-8                reprex_0.3.0            \n",
       " [70] base64enc_0.1-3          ggridges_0.5.2           png_0.1-7               \n",
       " [73] viridisLite_0.3.0        bitops_1.0-6             KernSmooth_2.23-17      \n",
       " [76] Biostrings_2.50.2        blob_1.2.1               qvalue_2.14.1           \n",
       " [79] parallelly_1.21.0        rstatix_0.6.0            gridGraphics_0.5-0      \n",
       " [82] ggsignif_0.6.0           scales_1.1.1             lpSolve_5.6.15          \n",
       " [85] memoise_1.1.0            ica_1.0-2                gdata_2.18.0            \n",
       " [88] zlibbioc_1.28.0          compiler_3.5.1           RColorBrewer_1.1-2      \n",
       " [91] fitdistrplus_1.1-1       cli_2.1.0                Rsamtools_1.34.1        \n",
       " [94] XVector_0.22.0           listenv_0.8.0            ps_1.4.0                \n",
       " [97] pbapply_1.4-3            htmlTable_2.0.1          Formula_1.2-3           \n",
       "[100] tidyselect_1.1.0         stringi_1.5.3            GOSemSim_2.8.0          \n",
       "[103] latticeExtra_0.6-28      fastmatch_1.1-0          tools_3.5.1             \n",
       "[106] future.apply_1.6.0       rio_0.5.16               rstudioapi_0.11         \n",
       "[109] uuid_0.1-4               foreign_0.8-76           farver_2.0.3            \n",
       "[112] Rtsne_0.15               ggraph_2.0.2             digest_0.6.27           \n",
       "[115] rvcheck_0.1.8            BiocManager_1.30.10      proto_1.0.0             \n",
       "[118] quadprog_1.5-8           Rcpp_1.0.5               car_3.0-9               \n",
       "[121] broom_0.7.0              RcppAnnoy_0.0.16         httr_1.4.2              \n",
       "[124] colorspace_1.4-1         rvest_0.3.5              fs_1.4.2                \n",
       "[127] XML_3.99-0.3             splines_3.5.1            uwot_0.1.8.9001         \n",
       "[130] expm_0.999-4             graphlayouts_0.7.0       ggplotify_0.0.5         \n",
       "[133] plotly_4.9.2.1           jsonlite_1.7.1           tidygraph_1.2.0         \n",
       "[136] UpSetR_1.4.0             R6_2.5.0                 Hmisc_4.4-0             \n",
       "[139] gsubfn_0.7               pillar_1.4.6             htmltools_0.5.0         \n",
       "[142] glue_1.4.2               codetools_0.2-16         fgsea_1.8.0             \n",
       "[145] tsne_0.1-3               sqldf_0.4-11             curl_4.3                \n",
       "[148] leiden_0.3.4             gtools_3.8.2             zip_2.0.4               \n",
       "[151] GO.db_3.7.0              openxlsx_4.2.2           survival_3.2-3          \n",
       "[154] munsell_0.5.0            DO.db_2.9                GenomeInfoDbData_1.2.0  \n",
       "[157] haven_2.3.1              reshape2_1.4.4           gtable_0.3.0            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
